docs/testing_servers.md

```diff
@@ -0,0 +1,340 @@
+    build a conformance payload
+
+    set the response data into the conformance payload
+
+    if this is the first response being sent then
+      set the request info into the conformance payload
+ 
+    sleep for any specified response delay
+
+    send the response
+
+  if an error was specified in the response definition then
+    if no responses have been sent yet
+      build an error with the specified error
+
+      set the following into the error details
+        * the received request
+        * any request headers
+```
+
+For the full documentation on handling the `ServerStream` endpoint, click [here][serverstream].
+
+### BidiStream
+
+The `BidiStream` endpoint is a bidirectional-streaming operation. It accepts 
+one-to-many requests of type `BidiStreamRequest` and returns one-to-many 
+responses of type `BidiStreamResponse`. 
+
+Similar to `ServerStream`, servers should immediately send response headers on 
+the stream before sleeping for any specified response delay and/or sending the 
+first message so that clients can be unblocked reading response headers.
+  
+If a response definition is not specified OR is specified, but response data
+is empty, the server should skip sending anything on the stream. When there
+are no responses to send, servers should throw an error if one is provided
+and return without error if one is not. Stream headers and trailers should
+still be set on the stream if provided regardless of whether a response is
+sent or an error is thrown.
+
+The `BidiStreamRequest` type specifies whether the operation should be full duplex
+or half duplex via the `full_duplex` field.
+
+If the `full_duplex` field is true:
+
+* the handler should read one request and then send back one response, and
+  then alternate, reading another request and then sending back another response, etc.
+  
+* if the server receives a request and has no responses to send, it
+  should throw the error specified in the request.
+  
+* the service should echo back all request properties in the first response
+  including the last received request. Subsequent responses should only
+  echo back the last received request.
+  
+* if the `response_delay_ms` duration is specified, the server should wait the given
+  duration after reading the request before sending the corresponding
+  response.
+  
+If the `full_duplex` field is false:
+
+* the handler should read all requests until the client is done sending.
+  Once all requests are read, the server should then send back any responses
+  specified in the response definition.
+  
+* the server should echo back all request properties, including all request
+  messages in the order they were received, in the first response. Subsequent
+  responses should only include the message data in the data field.
+  
+* if the `response_delay_ms` duration is specified, the server should wait that
+  long in between sending each response message.
+
+**Pseudocode**
+
+```text
```

Comment:
I think this might be easier to follow if we split it into two -- pseudo-code for the full-duplex logic and separate pseudo-code for the half-duplex logic.